### PR TITLE
Performance work

### PR DESCRIPTION
**Pull request by petersteneteg**
Core: Major refactorings.
* Serialization no longer supports use of the "allowReferences" to serialize pointer to the same object multiple times. Reasoning being that it requires all properties and ports, to always be present in the xml even if they don't have any state that needs serialization. The new solution uses "paths" (a dot seperated list of identifiers) instead of pointers where needed.
* Some of the serialization functions now supprorts filters and projections, to void having to create temporary objects to serialize.
* Property now has a `virtual bool isDefaultState() const` function. Should generally be overriden by subclases
* Property now has a `virtual bool needsDerialization() const` function, the default impl returns true, when the serliazation mode is All, or Default and !isDefaultState(), otherwise false.
* PropertyOwner now only serializes property that "needsDeserialization" this reduces.
* DefaultValues now uses StaticString to make the implementaion constexpr
* Many classIdentifiers now return const std::string& instead of std::string to avoud creating new string. This can easly be done by keeping a static std::string around.
* Now uses std::string_view instead of const std::string& in many places to avoid haing to create strings with dynamic allocations.
* StringConversion.h now has a StrBuffer class to make it easyier to format string using a stack buffer. Very useful for concatenation.
* StringConversion.h now has a forEachStringPart function to call a callback on parts of a string.
* Processor identifiers now follow the same rules as other identifiers
* MinMaxProperty is no longer a template property
* Animation update to core serialization changes
* Adds a compile time owning string class
* Adds a utility to convert a string_view to a c_str using a stack buffer
OpenGL: use std::string_view and StrBuffer to avoid creating strings

### CODE REVIEW

File: ext/ticpp/include/ticpp/ticpp.h [hunk 1]
```diff
@@ -206,7 +207,7 @@ namespace ticpp
         std::string ToString( const double& value ) const
         {
             std::stringstream convert;
-            convert.precision(40);
+            convert.precision(17);
```
Comment (ResearchDaniel) [inline ext/ticpp/include/ticpp/ticpp.h:210]: Is 17 enough for double precision?
Might consider adding a reference from where the constant is coming from.

File: include/inviwo/core/io/serialization/deserializer.h [hunk 2]
```diff
@@ -64,7 +66,7 @@ class IVW_CORE_API Deserializer : public SerializeBase, public LogFilter {
      * @param fileName path to file that is to be deserialized.
      * @param allowReference flag to manage references to avoid multiple object creation.
```
Comment (ResearchDaniel) [inline include/inviwo/core/io/serialization/deserializer.h]: remove obsolete comment

File: include/inviwo/core/io/serialization/deserializer.h [hunk 3]
```diff
@@ -74,7 +76,7 @@ class IVW_CORE_API Deserializer : public SerializeBase, public LogFilter {
      * deserialization.
      * @param allowReference flag to manage references to avoid multiple object creation.
```
Comment (ResearchDaniel) [inline include/inviwo/core/io/serialization/deserializer.h]: replace with refPath and update comment

File: include/inviwo/core/io/serialization/deserializer.h [hunk 4]
```diff
@@ -280,40 +287,53 @@ class IVW_CORE_API Deserializer : public SerializeBase, public LogFilter {
     int getInviwoWorkspaceVersion() const;
 
 private:
+    template <typename T>
+    using isDeserializable = decltype(std::declval<Deserializer>().deserialize(
+        std::declval<std::string_view>(), std::declval<T&>()));
+
+    template <typename T>
+    using isStreamable = decltype(std::declval<std::istream&>() >> std::declval<T&>());
```
Comment (ResearchDaniel) [inline include/inviwo/core/io/serialization/deserializer.h]: What does streamable mean?

File: include/inviwo/core/io/serialization/deserializer.h [hunk 5]
```diff
@@ -280,40 +287,53 @@ class IVW_CORE_API Deserializer : public SerializeBase, public LogFilter {
...
+
+    template <typename T>
+    static constexpr bool canDeserialize() {
+        return util::is_detected_exact_v<std::istream&, isStreamable, T> ||
+               util::is_detected_v<isDeserializable, T> || std::is_enum_v<T>;
+    }
+
+    static_assert(canDeserialize<float>());
```
Comment (ResearchDaniel) [inline include/inviwo/core/io/serialization/deserializer.h]: Are these for testing?

File: include/inviwo/core/metadata/metadata.h [hunk 6]
```diff
@@ -306,13 +306,14 @@ class MetaDataPrimitiveType<T, N, N> : public MetaData {
     MetaDataPrimitiveType(const MetaDataPrimitiveType& rhs);
     MetaDataPrimitiveType& operator=(const MetaDataPrimitiveType& that);
     virtual ~MetaDataPrimitiveType(){};
-    virtual std::string getClassIdentifier() const;
-    virtual MetaDataPrimitiveType<T, N, N>* clone() const;
-    virtual void serialize(Serializer& s) const;
-    virtual void deserialize(Deserializer& d);
+    virtual const std::string& getClassIdentifier() const override;
+    virtual MetaDataPrimitiveType<T, N, N>* clone() const override;
+    ;
```
Comment (ResearchDaniel) [inline include/inviwo/core/metadata/metadata.h]: remove

File: include/inviwo/core/network/networkedge.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace inviwo {
+
+class PropertyLink;
+class PortConnection;
+class ProcessorNetwork;
+
+struct IVW_CORE_API NetworkEdge : Serializable {
+    NetworkEdge() = default;
+    NetworkEdge(std::string asrc, std::string adst);
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/networkedge.h]: ?
I guess these are the identifier of source and destination port/property?

File: include/inviwo/core/network/networkedge.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    NetworkEdge(const PropertyLink& link);
+    NetworkEdge(const PortConnection& connection);
+
+    virtual ~NetworkEdge() = default;
+
+    PortConnection toConnection(const ProcessorNetwork& net) const;
+    PropertyLink toLink(const ProcessorNetwork& net) const;
+
+    void updateProcessorID(const std::unordered_map<std::string, std::string>& map);
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/networkedge.h:63]: Why and when should this be called?
What's in map?

File: include/inviwo/core/network/networkedge.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    virtual void serialize(Serializer& s) const override;
+    virtual void deserialize(Deserializer& d) override;
+
+    std::string src;
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/networkedge.h]: Are there restrictions and intents on the content of these? (reverse DNS corresponding to port/property class identifier)

File: include/inviwo/core/network/processornetwork.h [hunk 10]
```diff
@@ -226,6 +228,11 @@ class IVW_CORE_API ProcessorNetwork : public Serializable,
     std::vector<PropertyLink> getLinksBetweenProcessors(Processor* p1, Processor* p2);
 
     Property* getProperty(std::vector<std::string> path) const;
+    Property* getProperty(std::string_view path) const;
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/processornetwork.h:236]: We should document how path is represented. 
Path is a dot separated starting with the processor identifier (Processor::getIdentifier), e.g. processorid.propertyid

File: include/inviwo/core/network/processornetworkconverter.h [hunk 11]
```diff
@@ -65,6 +65,7 @@ class IVW_CORE_API ProcessorNetworkConverter : public VersionConverter {
     void updateProcessorIdentifiersStriped(TxElement* node);
     void updatePropertyEditorMetadata(TxElement* node);
     void updateCameraPropertyToRefs(TxElement* node);
+    void updateLinkAndConnections(TxElement* node);
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/processornetworkconverter.h:68]: Should we point to example usage in the class documentation? 
It is quite hard to understand how to use this class without examples.

File: include/inviwo/core/ports/port.h [hunk 12]
```diff
@@ -54,10 +54,12 @@ class IVW_CORE_API Port : public Serializable {
 
 public:
     virtual ~Port() = default;
-    std::string getIdentifier() const;
+    const std::string& getIdentifier() const;
     void setIdentifier(const std::string& name);
     Processor* getProcessor() const;
 
+    std::string getPathStr() const;
```
Comment (ResearchDaniel) [inline include/inviwo/core/ports/port.h]: Document. 
Should we default to usage of string for paths or std::vector<std::string>? If so, we could name this getPath?

File: include/inviwo/core/properties/buttonproperty.h [hunk 13]
```diff
@@ -87,6 +87,8 @@ class IVW_CORE_API ButtonProperty : public Property {
     // Override Property::resetToDefaultState, to avoid calling propertyModified  on reset.
     virtual ButtonProperty& resetToDefaultState() override;
 
+    virtual bool isDefaultState() const override;
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/buttonproperty.h:90]: Copydoc on all the overrides of isDefaultState?

File: include/inviwo/core/properties/compositeproperty.h [hunk 14]
```diff
@@ -74,6 +74,12 @@ class IVW_CORE_API CompositeProperty : public Property,
 
     virtual CompositeProperty& setCurrentStateAsDefault() override;
     virtual CompositeProperty& resetToDefaultState() override;
+
+    virtual bool isDefaultState() const override;
+
+    virtual bool needsSerialization() const override;
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/compositeproperty.h:83]: Copydoc on all needsSerialization?

File: include/inviwo/core/io/serialization/deserializer.h (hunk 2 above)
Comment (martinfalk) [inline include/inviwo/core/io/serialization/deserializer.h]: remove line referring to `allowReference`

File: include/inviwo/core/io/serialization/deserializer.h (hunk 3 above)
Comment (martinfalk) [inline include/inviwo/core/io/serialization/deserializer.h]: remove line `allowReference`

File: include/inviwo/core/io/serialization/deserializer.h (hunk 4 above)
Comment (martinfalk) [inline include/inviwo/core/io/serialization/deserializer.h]: is the in direction sufficient?

File: include/inviwo/core/io/serialization/deserializer.h [hunk 15]
```diff
@@ -280,40 +287,53 @@ class IVW_CORE_API Deserializer : public SerializeBase, public LogFilter {
...
+    static_assert(canDeserialize<int>());
+    static_assert(canDeserialize<std::string>());
```
Comment (martinfalk) [inline include/inviwo/core/io/serialization/deserializer.h]: why not test other things, too? enums, ifstream,...

File: include/inviwo/core/io/serialization/deserializer.h [hunk 16]
```diff
@@ -971,10 +1012,10 @@ void Deserializer::deserialize(const std::string& key, std::array<T, N>& cont,
 }
 
 template <typename K, typename V, typename C, typename A>
-void Deserializer::deserialize(const std::string& key, std::map<K, V, C, A>& map,
-                               const std::string& itemKey, const std::string& comparisonAttribute) {
-    if (!isPrimitiveType(typeid(K)) || comparisonAttribute.empty())
-        throw SerializationException("Error: map key has to be a primitive type", IVW_CONTEXT);
+void Deserializer::deserialize(std::string_view key, std::map<K, V, C, A>& map,
+                               std::string_view itemKey, std::string_view comparisonAttribute) {
+    static_assert(isPrimitiveType<K>(), "Error: map key has to be a primitive type");
```
Comment (martinfalk) [inline include/inviwo/core/io/serialization/deserializer.h:1008]: neat to have a compile-time check here.

File: include/inviwo/core/properties/minmaxproperty.h [hunk 17]
```diff
@@ -42,9 +44,9 @@ namespace inviwo {
  * A property representing a range.
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/minmaxproperty.h:44]: Not part of this PR, but might be worth updating:
We say that the property represents a range, which should mean that the value is a range (indeed, it is). But our functions "setRange" sets the bounds of the values, i.e. the range.
I think that it is unnessesarily confusing and could be  fixed by renaming range to bounds or min/max

File: include/inviwo/core/properties/property.h [hunk 18]
```diff
@@ -244,6 +249,19 @@ class IVW_CORE_API Property : public PropertyObservable,
...
 
+    /**
+     * Check if the property is in it's default state, i.e. resetToDefaultState would do nothing
+     * @see setCurrentStateAsDefault @see resetToDefaultState
+    */
+    virtual bool isDefaultState() const;
+
+    /**
+     * Determinate if the property should be included in the serialization
+     * Depends on the serialization mode and if the property is in the default state
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/property.h]: Would be useful to state how it (should) depend

File: include/inviwo/core/io/serialization/serializer.h [hunk 19]
```diff
@@ -61,14 +61,14 @@ class IVW_CORE_API Serializer : public SerializeBase {
...
-    Serializer(const std::string& fileName, bool allowReference = true);
+    Serializer(std::string_view fileName);
 
     virtual ~Serializer();
 
     /**
      * \brief Writes serialized data to the file specified by the currently set file name.
      *
-     * @note File name needs to be set before calling this method.
+     * @param file File to be written to.
```
Comment (martinfalk) [inline include/inviwo/core/io/serialization/serializer.h]: there is no argument `file`

File: include/inviwo/core/network/networkedge.h (hunk 7 above)
Comment (martinfalk) [inline include/inviwo/core/network/networkedge.h]: `asrc` and `adst`, why the `a` in the beginning?

File: include/inviwo/core/properties/optionproperty.h [hunk 20]
```diff
@@ -249,9 +250,12 @@ template <typename T>
 struct PropertyTraits<TemplateOptionProperty<T>> {
     static std::string classIdentifier() {
         if constexpr (std::is_enum_v<T>) {
-            return "org.inviwo.OptionProperty" + util::enumName<T>();
+            static const std::string identifier = "org.inviwo.OptionProperty" + util::enumName<T>();
```
Comment (martinfalk) [inline include/inviwo/core/properties/optionproperty.h:253]: enumName returns a std::string ;-)

File: include/inviwo/core/util/defaultvalues.h [hunk 21]
```diff
@@ -32,138 +32,186 @@
...
+    StaticString<N> name;
     uvec2 dim;
     T val;
     T min;
     T max;
     T inc;
 };
 
+template <typename T, size_t N>
+InviwoDefaultData(StaticString<N>, uvec2, T, T, T, T) -> InviwoDefaultData<T, N>;
```
Comment (martinfalk) [inline include/inviwo/core/util/defaultvalues.h]: reason for this? Is this a hint to always use the constructor of the non-specialized template?

File: include/inviwo/core/util/introspection.h [hunk 22]
```diff
@@ -82,13 +87,13 @@ template <class T>
...
 template <typename T>
-std::string classIdentifier() {
+const std::string& classIdentifier() {
     if constexpr (HasClassIdentifierUpper<T>::value) {
         return T::CLASS_IDENTIFIER;
     } else if constexpr (HasClassIdentifierLower<T>::value) {
         return T::classIdentifier;
     } else {
-        return {};
+        static_assert(detail::alwaysFalse<T>, "ClassIdentifier is missing for type");
```
Comment (martinfalk) [inline include/inviwo/core/util/introspection.h:96]: don't you need to call `alwaysFalse<T>()` here?

File: include/inviwo/core/util/stdextensions.h [hunk 23]
```diff
@@ -139,7 +139,40 @@ struct overloaded : Ts... {
     using Ts::operator()...;
 };
 template <class... Ts>
-overloaded(Ts...)->overloaded<Ts...>;
+overloaded(Ts...) -> overloaded<Ts...>;
+
+struct identity {
+    template <typename T>
+    constexpr decltype(auto) operator()(T&& t) const noexcept {
```
Comment (martinfalk) [inline include/inviwo/core/util/stdextensions.h:146]: need for following overloads instead?
```c++
constexpr decltype(auto) operator()(T&& t) const&
```
and
```
constexpr decltype(auto) operator()(T&& t) &&
```

Comment (martinfalk) [reply include/inviwo/core/util/stdextensions.h:146]: see also talk by Nicolai Josuttis https://youtu.be/TFMKjL38xAI?t=2813

File: modules/animation/tests/unittests/track-test.cpp [hunk 24]
```diff
@@ -277,6 +277,8 @@ TEST(AnimationTests, KeyframeSequenceSerializationTest) {
 }
 
 TEST(AnimationTests, TrackSerializationTest) {
+    /*
```
Comment (martinfalk) [inline modules/animation/tests/unittests/track-test.cpp]: disabled on purpose?

File: modules/animation/tests/unittests/track-test.cpp [hunk 25]
```diff
@@ -332,9 +334,12 @@ TEST(AnimationTests, TrackSerializationTest) {
     interpolationFactory.unRegisterObject(&linearIFO);
     propertyFactory.unRegisterObject(&floatPFO);
     trackFactory.unRegisterObject(&floatTFO);
+
+    */
 }
 
 TEST(AnimationTests, AnimationSerializationTest) {
+    /*
```
Comment (martinfalk) [inline modules/animation/tests/unittests/track-test.cpp]: disabled on purpose?

File: modules/opengl/include/modules/opengl/shader/shadermanager.h [hunk 26]
```diff
@@ -114,6 +116,4 @@ std::shared_ptr<ShaderManager::Callback> ShaderManager::onWillRemoveShader(T&& c
     return shaderRemoveCallbacks_.add(std::forward<T>(callback));
 }
 
-}  // namespace inviwo
-
-#endif  // IVW_SHADERMANAGER_H
+}  // namespace inviwo
```
Comment (martinfalk) [inline modules/opengl/include/modules/opengl/shader/shadermanager.h]: missing newline

File: modules/opengl/src/shader/shaderobject.cpp [hunk 27]
```diff
@@ -536,24 +535,22 @@ void ShaderObject::parseSource(std::ostringstream& output) {
...
 
-    std::string curLine;
-    while (std::getline(origShaderInfoLog, curLine)) {
+    util::forEachStringPart(compileLog, "\n", [&](std::string_view curLine) {
         if (!curLine.empty()) {
             const int origLineNumber = utilgl::getLogLineNumber(curLine);
             if (origLineNumber > 0) {
-                const auto res = lnr_.resolveLine(origLineNumber);
+                const auto& res = lnr_.resolveLine(origLineNumber);
                 result << "\n"
```
Comment (martinfalk) [inline modules/opengl/src/shader/shaderobject.cpp]: why using a stringstream and not fmt::format here?

File: modules/opengl/src/shader/shaderobject.cpp [hunk 28]
```diff
@@ -687,44 +701,41 @@ std::string ShaderObject::print(bool showSource, bool showPreprocess) {
...
-                const std::string file =
-                    res.first.empty() ? "" : splitString(res.first, '/').back();
-                const size_t lineNumber = res.second;
+            util::forEachStringPart(sourceProcessed_, "\n", [&](std::string_view line) {
+                const auto& res = lnr_.resolveLine(i);
+                const auto file = res.first.empty() ? std::string_view{""}
+                                                    : splitStringView(res.first, '/').back();
+                const auto lineNumber = res.second;
 
                 out << std::left << std::setw(width + 1u) << file << std::right << std::setw(4)
```
Comment (martinfalk) [inline modules/opengl/src/shader/shaderobject.cpp:721]: fmt::format?

File: modules/opengl/src/shader/shaderobject.cpp [hunk 29]
```diff
@@ -687,44 +701,41 @@ std::string ShaderObject::print(bool showSource, bool showPreprocess) {
...
             size_t lineNumber = 1;
-            std::string line;
             std::stringstream out;
-            std::istringstream in(resource_->source());
-            while (std::getline(in, line)) {
-                std::string file = resource_->key();
+            util::forEachStringPart(resource_->source(), "\n", [&](std::string_view line) {
+                const auto& file = resource_->key();
 
                 out << std::left << std::setw(file.length() + 1u) << file << std::right
```
Comment (martinfalk) [inline modules/opengl/src/shader/shaderobject.cpp:736]: fmt::format?

File: modules/opengl/src/shader/shaderutils.cpp [hunk 30]
```diff
@@ -501,50 +503,50 @@ void addShaderDefinesBGPort(Shader& shader, const ImageInport& port) {
...
-        planes += indicator.plane3_ ? ", params.plane3" : "";
-        value =
-            "drawPlanes(result, samplePosition, rayDirection, increment " + planes + ",t,tDepth)";
-
-        shader.getFragmentShaderObject()->addShaderDefine("PLANES_ENABLED");
+    constexpr std::string_view key =
+        "DRAW_PLANES(result, samplePosition, rayDirection, increment, params,t,tDepth)";
+    if (enabled) {
+        StrBuffer planes(
+            "drawPlanes(result, samplePosition, rayDirection, increment {}{}{}, t, tDepth)",
```
Comment (martinfalk) [inline modules/opengl/src/shader/shaderutils.cpp]: remove space after `increment`

File: include/inviwo/core/properties/property.h [hunk 31]
```diff
@@ -74,7 +74,7 @@ struct PropertyTraits {
      * "T::CLASS_IDENTIFIER". In case it is not found an empty string will be returned. An empty
      * class identifier will be considered an error in various factories.
      */
-    static std::string classIdentifier() { return util::classIdentifier<T>(); }
+    static const std::string& classIdentifier() { return util::classIdentifier<T>(); }
```
Comment (martinfalk) [inline include/inviwo/core/properties/property.h:80]: should be documented in the changelog

File: include/inviwo/core/properties/property.h [hunk 32]
```diff
@@ -244,6 +249,19 @@ class IVW_CORE_API Property : public PropertyObservable,
      */
     virtual Property& resetToDefaultState();
 
+    /**
+     * Check if the property is in it's default state, i.e. resetToDefaultState would do nothing
+     * @see setCurrentStateAsDefault @see resetToDefaultState
+    */
+    virtual bool isDefaultState() const;
```
Comment (martinfalk) [inline include/inviwo/core/properties/property.h:263]: should maybe documented in the changelog

File: include/inviwo/core/properties/property.h [hunk 33]
```diff
@@ -244,6 +249,19 @@ class IVW_CORE_API Property : public PropertyObservable,
...
+
+    /**
+     * Determinate if the property should be included in the serialization
+     * Depends on the serialization mode and if the property is in the default state
+     */
+    virtual bool needsSerialization() const;
```
Comment (martinfalk) [inline include/inviwo/core/properties/property.h:271]: should maybe documented in the changelog

File: src/core/properties/cameraproperty.cpp [hunk 34]
```diff
@@ -44,6 +44,8 @@
 
 #include <limits>
 
+#include <tracy/Tracy.hpp>
```
Comment (martinfalk) [inline src/core/properties/cameraproperty.cpp]: remove!

File: src/core/properties/cameraproperty.cpp [hunk 35]
```diff
@@ -376,11 +385,13 @@ void CameraProperty::addCamerapProperty(std::unique_ptr<Property> camprop) {
 }
 
 void CameraProperty::serialize(Serializer& s) const {
+    ZoneScopedNC("CameraProperty serialize", 0xAA0000);
     CompositeProperty::serialize(s);
     s.serialize("Camera", camera_);
 }
 
 void CameraProperty::deserialize(Deserializer& d) {
+    ZoneScopedNC("CameraProperty deserialize", 0xAA0000);
```
Comment (martinfalk) [inline src/core/properties/cameraproperty.cpp]: tracy...

File: src/core/properties/cameraproperty.cpp [hunk 36]
```diff
@@ -376,11 +385,13 @@ void CameraProperty::addCamerapProperty(std::unique_ptr<Property> camprop) {
 }
 
 void CameraProperty::serialize(Serializer& s) const {
+    ZoneScopedNC("CameraProperty serialize", 0xAA0000);
```
Comment (martinfalk) [inline src/core/properties/cameraproperty.cpp]: tracy...

File: src/core/tests/benchmarks/CMakeLists.txt [hunk 37]
```diff
@@ -0,0 +1,21 @@
...
+find_package(benchmark CONFIG REQUIRED)
+target_link_libraries(bm-safecstr 
+    PUBLIC 
+        benchmark::benchmark
+        inviwo::core
+)
+set_target_properties(bm-safecstr PROPERTIES FOLDER benchmarks)
+
+set_property(TARGET bm-safecstr APPEND_STRING PROPERTY LINK_FLAGS 
+   " /SUBSYSTEM:CONSOLE /ENTRY:mainCRTStartup")
```
Comment (martinfalk) [inline src/core/tests/benchmarks/CMakeLists.txt]: looks very MSVC specific

**Review by martinfalk — CHANGES_REQUESTED** [review]
nice work. Code looks much cleaner now in places.

File: include/inviwo/core/util/rendercontext.h [hunk 38]
```diff
@@ -40,6 +40,7 @@ namespace inviwo {
  * An abstraction for what we need from a "context".
  */
 class IVW_CORE_API ContextHolder {
+
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/rendercontext.h]: remove

File: include/inviwo/core/util/safecstr.h [hunk 39]
```diff
@@ -0,0 +1,61 @@
...
+#pragma once
+
+#include <string_view>
+#include <array>
+#include <memory>
+
+namespace inviwo {
+
+template <size_t N = 120>
+class SafeCStr {
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/safecstr.h:45]: Document. What does Safe mean?

File: include/inviwo/core/util/staticstring.h [hunk 40]
```diff
@@ -0,0 +1,168 @@
...
+template <size_t N>
+struct StaticString;
+
+namespace detail {
+
+// Not constexpr in 17
+template <class InputIt, class Size, class OutputIt>
+constexpr OutputIt copy_n(InputIt first, Size count, OutputIt result) {
+    if (count > 0) {
+        *result++ = *first;
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/staticstring.h:49]: Why not for all: 
*result++ = *first++; 
or, if above is not possible: 
*result++ = *first; first++; 
No big deal, but I think that it would be a bit cleaner

File: include/inviwo/core/util/stringconversion.h [hunk 41]
```diff
@@ -60,6 +62,34 @@ T stringTo(const std::string& str) {
     return result;
 }
 
+struct IVW_CORE_API StrBuffer {
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/stringconversion.h:70]: Document what and when to use and when not to use.

File: modules/animation/tests/unittests/track-test.cpp (hunk 24 above)
Comment (ResearchDaniel) [inline modules/animation/tests/unittests/track-test.cpp]: Remove or fix :)

File: modules/animation/tests/unittests/track-test.cpp (hunk 25 above)
Comment (ResearchDaniel) [inline modules/animation/tests/unittests/track-test.cpp]: Remove or fix

File: include/inviwo/core/util/safecstr.h (hunk 39 above)
Comment (ResearchDaniel) [reply include/inviwo/core/util/safecstr.h:45]: You probably want to mention that strings of size < N are allocated on the stack while heap allocation is used otherwise.
It is not safe in the sense that out-of-memory will cause errors...

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
Sweet!
I have some minor comments that you might want to consider

File: ext/ticpp/include/ticpp/ticpp.h (hunk 1 above)
Comment (petersteneteg) [reply ext/ticpp/include/ticpp/ticpp.h:210]: Yes

File: include/inviwo/core/io/serialization/deserializer.h (hunk 4 above)
Comment (petersteneteg) [reply include/inviwo/core/io/serialization/deserializer.h]: this is only for the deserialisation

File: include/inviwo/core/io/serialization/deserializer.h (hunk 15 above)
Comment (petersteneteg) [reply include/inviwo/core/io/serialization/deserializer.h]: removed, was only for debugging

File: include/inviwo/core/network/networkedge.h (hunk 7 above)
Comment (petersteneteg) [reply include/inviwo/core/network/networkedge.h]: to avoid name hiding

File: include/inviwo/core/properties/optionproperty.h (hunk 20 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/optionproperty.h:253]: that's fine will only be called once

File: include/inviwo/core/util/defaultvalues.h (hunk 21 above)
Comment (petersteneteg) [reply include/inviwo/core/util/defaultvalues.h]: yes, a deduction guide

File: include/inviwo/core/util/introspection.h (hunk 22 above)
Comment (petersteneteg) [reply include/inviwo/core/util/introspection.h:96]: alwaysFalse is a variable not a function

File: include/inviwo/core/util/stdextensions.h (hunk 23 above)
Comment (petersteneteg) [reply include/inviwo/core/util/stdextensions.h:146]: can you explain why until I have to to look at that talk? don't understand why the value category of the function object should matter

File: modules/animation/tests/unittests/track-test.cpp (hunk 24 above)
Comment (petersteneteg) [reply modules/animation/tests/unittests/track-test.cpp]: enabled again

File: modules/animation/tests/unittests/track-test.cpp (hunk 25 above)
Comment (petersteneteg) [reply modules/animation/tests/unittests/track-test.cpp]: enabled

File: modules/opengl/src/shader/shaderobject.cpp (hunk 28 above)
Comment (petersteneteg) [reply modules/opengl/src/shader/shaderobject.cpp:721]: some other time

File: modules/opengl/src/shader/shaderobject.cpp (hunk 29 above)
Comment (petersteneteg) [reply modules/opengl/src/shader/shaderobject.cpp:736]: some other time

File: include/inviwo/core/properties/property.h (hunk 31 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/property.h:80]: actually since this is just a static function not virtual or any such things, it does not matter or break any thing.
your specialisation can return whatever...

File: include/inviwo/core/io/serialization/deserializer.h (hunk 4 above)
Comment (petersteneteg) [reply include/inviwo/core/io/serialization/deserializer.h]: that it can be read from a istream.

File: include/inviwo/core/io/serialization/deserializer.h (hunk 5 above)
Comment (petersteneteg) [reply include/inviwo/core/io/serialization/deserializer.h]: removed

File: include/inviwo/core/network/networkedge.h (hunk 9 above)
Comment (petersteneteg) [reply include/inviwo/core/network/networkedge.h]: this is just a list of period separated identifiers as usual.

File: include/inviwo/core/network/processornetworkconverter.h (hunk 11 above)
Comment (petersteneteg) [reply include/inviwo/core/network/processornetworkconverter.h:68]: just look at the code here for examples

File: include/inviwo/core/ports/port.h (hunk 12 above)
Comment (petersteneteg) [reply include/inviwo/core/ports/port.h]: changed getPath instead

File: include/inviwo/core/properties/compositeproperty.h (hunk 14 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/compositeproperty.h:83]: No https://www.doxygen.nl/manual/config.html#cfg_inherit_docs

File: include/inviwo/core/properties/buttonproperty.h (hunk 13 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/buttonproperty.h:90]: No https://www.doxygen.nl/manual/config.html#cfg_inherit_docs

File: include/inviwo/core/properties/minmaxproperty.h (hunk 17 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/minmaxproperty.h:44]: Agree, but this PR is already too big....

File: include/inviwo/core/util/staticstring.h (hunk 40 above)
Comment (petersteneteg) [reply include/inviwo/core/util/staticstring.h:49]: I just used the default impl from c++ reference, think the reason for the strange logic might come from http://www.open-std.org/jtc1/sc22/wg21/docs/lwg-active.html#2471
When we move to c++20 we std::copy_n is constexpr and this can be removed.

File: include/inviwo/core/util/stringconversion.h (hunk 41 above)
Comment (ResearchDaniel) [reply include/inviwo/core/util/stringconversion.h:70]: We might consider using memory-aligned size instead of 500. Not sure if it makes a difference, but 512?

**Review by ResearchDaniel — DISMISSED** [review]

File: include/inviwo/core/util/stringconversion.h (hunk 41 above)
Comment (petersteneteg) [reply include/inviwo/core/util/stringconversion.h:70]: Thats the default from fmt did not modify it my self...

File: include/inviwo/core/util/stringconversion.h [hunk 42]
```diff
@@ -155,6 +155,12 @@ constexpr void forEachStringPart(std::string_view str, std::string_view sep, Fun
     }
 }
 
+inline std::pair<std::string_view, std::string_view> divideBy(std::string_view str,
```
Comment (ResearchDaniel) [inline include/inviwo/core/util/stringconversion.h]: splitBy would be more consistent with the naming of splitString

Comment (petersteneteg) [reply include/inviwo/core/util/stringconversion.h]: True. But I considered that to be to similar, since splitString also takes a separator... I should add comment though...

Comment (petersteneteg) [reply include/inviwo/core/util/stringconversion.h]: renamed to splitByFirst and splitByLast

File: include/inviwo/core/network/processornetwork.h [hunk 43]
```diff
@@ -225,7 +227,38 @@ class IVW_CORE_API ProcessorNetwork : public Serializable,
     std::vector<Property*> getPropertiesLinkedTo(Property* property);
     std::vector<PropertyLink> getLinksBetweenProcessors(Processor* p1, Processor* p2);
 
-    Property* getProperty(std::vector<std::string> path) const;
+    /**
+     * @breif Get Property by path
```
Comment (martinfalk) [inline include/inviwo/core/network/processornetwork.h]: typo, should be `@brief`

File: include/inviwo/core/network/processornetwork.h [hunk 44]
```diff
@@ -225,7 +227,38 @@ class IVW_CORE_API ProcessorNetwork : public Serializable,
...
+     * @param path string of dot separated identifiers starting with a processor identifier followed
+     * by property identifiers.
+     * @return the property or nullptr if not found
+     */
+    Property* getProperty(std::string_view path) const;
+
+    /**
+     * @breif Get Port by path
```
Comment (martinfalk) [inline include/inviwo/core/network/processornetwork.h]: `@brief`

File: include/inviwo/core/network/processornetwork.h [hunk 45]
```diff
@@ -225,7 +227,38 @@ class IVW_CORE_API ProcessorNetwork : public Serializable,
...
+     * @param path string of dot separated identifiers starting with a processor identifier followed
+     * by a port identifier.
+     * @return the port or nullptr if not found
+     */
+    Port* getPort(std::string_view path) const;
+
+    /**
+     * @breif Get Inport by path
```
Comment (martinfalk) [inline include/inviwo/core/network/processornetwork.h]: `@brief`

File: include/inviwo/core/network/processornetwork.h [hunk 46]
```diff
@@ -225,7 +227,38 @@ class IVW_CORE_API ProcessorNetwork : public Serializable,
...
+     * @param path string of dot separated identifiers starting with a processor identifier followed
+     * by a port identifier.
+     * @return the port or nullptr if not found
+     */
+    Inport* getInport(std::string_view path) const;
+
+    /**
+     * @breif Get Outport by path
```
Comment (martinfalk) [inline include/inviwo/core/network/processornetwork.h]: `@brief`

File: include/inviwo/core/util/stringconversion.h [hunk 47]
```diff
@@ -80,9 +143,78 @@ IVW_CORE_API std::wstring toWstring(const std::string& str);
...
+
+/**
+ * @brief Divide a string into two parts by the first instance of a delimiter
+ * @param str string to divide
+ * @param delimiter not include in either returned strings
+ * @return a pair of strings, if the delimiter is not found the first string will be the same as the
+ * input str, and the second one will be empty
+ */
+constexpr std::pair<std::string_view, std::string_view> splitByFirst(std::string_view str,
+                                                                     char delimeter = ' ') {
```
Comment (martinfalk) [inline include/inviwo/core/util/stringconversion.h]: typo `delimiter`, (search&replace)

**Review by martinfalk — DISMISSED** [review]

**Review by martinfalk — APPROVED** [review]
